File: Data Structures & Algorithms/reorder-linked-list/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public void reorderList(ListNode head) {
        ListNode node = head;
        Map<Integer, ListNode> maps = new HashMap<>();
        int index = 0;
        while (node!=null) {
            maps.put(index, node);
            index += 1;
            node = node.next;
        }
        int left = 0;
        int right = index-1;
        ListNode curr = head;
        while (left < right) {
            ListNode temp = curr.next;
            curr.next = maps.get(right);
            curr = curr.next;
            curr.next = temp;
            curr = curr.next;
            left+= 1;
            right -= 1;
        }
        curr.next = null;

    }
}

File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:    
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def merge(l1, l2):
            res = ListNode(0)
            head = res
            while l1 and l2:
                if l1.val < l2.val:
                    res.next = l1
                    l1 = l1.next
                else:
                    res.next = l2
                    l2 = l2.next
                res = res.next
            
            while l1:
                res.next = l1
                l1 = l1.next
                res = res.next
            while l2:
                res.next = l2
                l2 = l2.next
                res = res.next
            return head.next
        
        def mergeLists(lists):
            if len(lists) == 1:
                return lists[0]
            if len(lists) == 0:
                return
            mid = len(lists) // 2
            left = mergeLists(lists[:mid])
            right = mergeLists(lists[mid:])
            return merge(left,right)
        
        return mergeLists(lists)
                
        